### PR TITLE
Fix State.SyncChanges mutex Lock/Unlock

### PR DESCRIPTION
**Pull request by PumpkinSeed**
I try to accomplish as minimal lists about the necessary fields as I could, so I created two function scope map.

### CODE REVIEW

**Comment by Aestek** [comment]
If I'm not mistaken this introduces a race for both service and checks.

When deleting a service :

1. `SyncChanges()` is called, the diff is computed, lock is released 
2. `deleteService()` is called and makes a `Catalog.Deregister` RPC
2. A service with the same id is registered 
3. The RPC returns and the service is removed from the map when it should not have been.

When updating a service : 

1. `SyncChanges()` is called, the diff is computed, lock is released 
2. `syncService()` is called and makes a `Catalog.Register` RPC
2. The service is updated
3. The RPC returns and the service is marked as "InSync" when its in fact not and the update is delayed until next anti-antropy run.

I'm not sure what the best way to fix this would be, but one way would be to maintain a state version number : the version is incremented on each state change and is stored in the ServiceState/CheckState structs. The `{sync,delete}{Service,Check}` funcs are called with the version their respecting entities had when the diff was computed (and the lock was held). If when the RPCs return the entity version is not what was passed as arguments the entity is not marked as in sync nor deleted from state.

Additionally, `delete{Service,Check}` currently expect the state lock to be held when called. This introduces a crash due to the concurrent writes to `State.{services,checks}`.

**Comment by PumpkinSeed** [comment]
@Aestek I see the complexity behind this problem. Let me have an other run with it.

**Comment by PumpkinSeed** [comment]
@Aestek I think it get fixed. So I'm checking the {ServiceState,CheckState}.Deleted, because in case of AddService these will be false, so after the RPC response arrives and the Deleted still true, it should mean there is no new service under the same id.

**Comment by Aestek** [comment]
@PumpkinSeed this fixes the "delete" scenario, but not the "update"

**Comment by PumpkinSeed** [comment]
@Aestek yes, it's a bit tricky. I would like to have a bit more time to investigate about.

**Comment by PumpkinSeed** [comment]
@Aestek I tried it with your suggestion of versioning.

**Comment by jacksontj** [comment]
I just ran into this same issue (actually ran into it because of disk space issues on the master nodes which caused API slowdown -- https://github.com/hashicorp/consul/issues/866). While debugging this I set up a repro environment locally (master + agent with tc adding 3s latency in the middle). I have pulled down the PR branch and am running this build and it seems to resolve the issue.

I've also read through the PR and it seems like a reasonable fix to the problem (versioning works since its single-source data -- from the agent -- that needs to eventually be pushed to the masters).

So I am very interested in getting this merged in, so if there is anything I can do to help this along let me know :)

**Comment by Aestek** [comment]
This looks good to me as well

**Comment by PumpkinSeed** [comment]
May I ask what is the status of the PR? I have seen there is conflict. I will check it out tonight.

**Comment by jacksontj** [comment]
@freddygv any update on this PR? This is a definite liability for users of consul agent services.

File: agent/local/state.go [hunk 1]
```diff
@@ -376,9 +402,11 @@ func (l *State) setServiceStateLocked(s *ServiceState) {
 	l.services[key] = s
 
 	if hasOld && old.WatchCh != nil {
+		l.services[structs.NewServiceID(s.Service.ID, nil)].version = old.version
```
Comment (i0rek) [inline agent/local/state.go]: This can go away since we are using equality to check the version.

```suggestion
```

File: agent/local/state.go [hunk 2]
```diff
@@ -1251,9 +1346,14 @@ func (l *State) syncCheck(key structs.CheckID) error {
 
 	var out struct{}
 	err := l.Delegate.RPC("Catalog.Register", &req, &out)
+	l.Lock()
+	defer l.Unlock()
 	switch {
 	case err == nil:
 		l.checks[key].InSync = true
```
Comment (i0rek) [inline agent/local/state.go]: We need to check if the key is in the map here as well.

```suggestion
                if c, ok := l.checks[key]; ok && c.version == currentVersion {
                  c.InSync = true
                }
```

File: agent/local/state.go [hunk 3]
```diff
@@ -1251,9 +1346,14 @@ func (l *State) syncCheck(key structs.CheckID) error {
...
+		if l.checks[key].version > currentVersion {
+			l.checks[key].InSync = false
+		}
 		// Given how the register API works, this info is also updated
 		// every time we sync a check.
 		l.nodeInfoInSync = true
```
Comment (i0rek) [inline agent/local/state.go]: We don't know it is in sync because it is not versioned. 

```suggestion
```

Comment (i0rek) [inline agent/local/state.go]: There is a case with `acl` stuff below, which also needs to check the key before setting `InSync`, because it could gone.

File: agent/local/state.go [hunk 4]
```diff
@@ -21,6 +21,7 @@ import (
 )
 
 const fullSyncReadMaxStale = 2 * time.Second
+const versionRollover = 65533
```
[the remark this answers is not in the document]
Comment (PumpkinSeed) [reply agent/local/state.go]: If it doesn't have to do a rollover then we don't need this const at all.

[the remark this answers is not in the document]
Comment (i0rek) [reply agent/local/state.go]: you are right.

**Comment by PumpkinSeed** [comment]
I fixed nearly all of the suggested things. But I already lost my domain knowledge since I created the PR and also it's started to get a hell of changes and I can't oversee it properly. 
So what do you think about a bit of restructure? @i0rek @freddygv 
Let me explain my suggestion:
- ServiceState and CheckState should get their own file line service_state.go and check_state.go
- They should have methods for changing their InSync, because there are more places where we want to call it, and I think it make sense to have a general solution which checks everything properly before we want to perform that. 
- I prefer to move the related State struct's methods along with their main struct. ex.: `func (l *State) CheckState(id structs.CheckID) *CheckState` goes to check_state.go

Then when I oversee the problem again properly we can get back to the bigger issue what you told me. Or if you don't want to have this restructure I can get back my domain on the current implementation. But I recommend that, because it's really hard to manage it right now I guess. 

Let me know what you think.

**Comment by freddygv** [comment]
@PumpkinSeed, after having a team-wide discussion today, we would like to suggest that our team makes the remaining changes to this PR. 

Since the anti-entropy sync is so fundamental in the operation of Consul, we need to spend a significant amount of additional time making sure that:
- Remaining corner-cases are addressed
- Everything works as expected
- There is appropriate test coverage

Due to the time commitment needed we also need to decide how to prioritize this work relative to other features on our roadmap. So the enhancement could end up on hold again because of that.

Lastly, please note that we would like to preserve the changes that you have made, and if we do, we will ensure that you are credited in the git history.

Thanks again for your contribution. We really appreciate the time and effort you have put into this problem.

**Comment by PumpkinSeed** [comment]
Probably this is the best for all of us. Because I'm totally out of context. Thank you for that help.

**Comment by jacksontj** [comment]
@freddygv any updated ETA on this fix? I've bee hit by this issue a couple of times and am anxious to get a fix in place.

**Comment by freddygv** [comment]
Hi @jacksontj, we picked this back up earlier this week. It should make its way into the next release as long as we feel comfortable with the changes.

**Comment by jacksontj** [comment]
@freddygv any update on this issue?

**Comment by freddygv** [comment]
Hi @jacksontj we have been working on and off on the PR, but didn't have much time due to the 1.8.0 release. Going to take another pass at it next week and should have a more concrete update then.

**Comment by jacksontj** [comment]
@freddygv just checking in again here, really excited to get this fix in :)

**Comment by freddygv** [comment]
@jacksontj we're still working on an implementation on a new branch. I'll post here when we have a PRs out.

**Comment by jacksontj** [comment]
@freddygv checking in now thats it been another 8 months :)

**Comment by i0rek** [comment]
@jacksontj thanks for reaching out! We worked on this change quite a bit but don't have anything to show so far. This is a really challenging task because it changes Consul behaviour significantly. And while we still think this is a great idea, we no longer think it can be easily done. 
We decided to stop working on this until we have the time to properly think about the implications of this change in advance and then decided on how to implement it so that it doesn't break the current behaviour.
